### PR TITLE
Port to 1.4

### PR DESCRIPTION
**Pull request by Pop000100**
See commit descriptions for a changelogs.

I am going to wait a day to merge this.

### CODE REVIEW

File: NPCs/BaseballCaveBat.cs [hunk 1]
```diff
@@ -38,16 +38,12 @@ public override float SpawnChance(NPCSpawnInfo spawnInfo)
...
-			if(Main.rand.Next(250) == 0)
-				Item.NewItem(npc.position, npc.width, npc.height, ItemID.ChainKnife, prefixGiven: -1);
-			if(Main.rand.Next(100) == 0)
-				Item.NewItem(npc.position, npc.width, npc.height, ItemID.DepthMeter, prefixGiven: -1);
-			if(Main.rand.Next(Main.expertMode ? 2 : 4) == 0)
-				Item.NewItem(npc.position, npc.width, npc.height, ModContent.ItemType<BaseballBat>(), prefixGiven: -1);
-			if(Main.rand.Next(3) == 0)
-				Item.NewItem(npc.position, npc.width, npc.height, ModContent.ItemType<BaseballCap>());
+			npcLoot.Add(ItemDropRule.Common(ItemID.ChainKnife, 250));
+			npcLoot.Add(ItemDropRule.Common(ItemID.DepthMeter, 100));
```
Comment (gardenappl) [inline NPCs/BaseballCaveBat.cs:44]: Do we need to add `ItemDropRule`s for vanilla items? I honsetly don't remember why I had that code in the first place.

File: NPCs/BaseballCaveBat.cs [hunk 2]
```diff
@@ -38,16 +38,12 @@ public override float SpawnChance(NPCSpawnInfo spawnInfo)
...
+			npcLoot.Add(ItemDropRule.NormalvsExpert(ModContent.ItemType<BaseballBat>(), 4, 2));
+			npcLoot.Add(ItemDropRule.Common(ModContent.ItemType<BaseballCap>(), 3));
```
Comment (gardenappl) [inline NPCs/BaseballCaveBat.cs:46]: Is that 50% drop rate in Expert mode? What was I thinking? Maybe I intended this to be 2%, 4%, and 3%.

Comment (gardenappl) [reply NPCs/BaseballCaveBat.cs:46]: Then we'd have to swap the NormalvsExpert rule so it's 2% in normal mode and 4% in expert mode.

File: NPCs/BaseballCaveBat.cs (hunk 1 above)
Comment (gardenappl) [reply NPCs/BaseballCaveBat.cs:44]: Wait, I'm dumb, I forgot that these are custom entities and not `GlobalNPC`s, haha.

File: NPCs/BaseballCaveBat.cs (hunk 2 above)
Comment (gardenappl) [reply NPCs/BaseballCaveBat.cs:46]: (whoops, haha)

File: Projectiles/AutofisherBobber.cs [hunk 3]
```diff
@@ -51,200 +52,201 @@ public override void SetDefaults()
...
-				//float num5 = (float)(Main.player[projectile.owner].position.X + (double)(Main.player[projectile.owner].width / 2) - vector2.X);
-				//float num6 = (float)(Main.player[projectile.owner].position.Y + (double)(Main.player[projectile.owner].height / 2) - vector2.Y);
+				//if((double)Projectile.localAI[0] < 100.0)
+				//	++Projectile.localAI[0];
+				Projectile.tileCollide = false;
+				const double productDivisor = 15.8999996185303;
+				const double ten = 10;
+				Vector2 vector2 = new Vector2((float)(Projectile.position.X + (double)Projectile.width * 0.5), (float)(Projectile.position.Y + (double)Projectile.height * 0.5));
+				//float num5 = (float)(Main.player[Projectile.owner].position.X + (double)(Main.player[Projectile.owner].width / 2) - vector2.X);
+				//float num6 = (float)(Main.player[Projectile.owner].position.Y + (double)(Main.player[Projectile.owner].height / 2) - vector2.Y);
```
Comment (gardenappl) [inline Projectiles/AutofisherBobber.cs]: I would've just removed the commented-out code if I were you.😅 But this is fine, too.

File: Projectiles/AutofisherBobber.cs [hunk 4]
```diff
@@ -51,200 +52,201 @@ public override void SetDefaults()
...
 					if(((Rectangle)rectangle1).Intersects(rectangle2))
 					{
-						if((double)projectile.ai[0] > 0.0)
+						if((double)Projectile.ai[0] > 0.0)
 						{
-							int Type = (int)projectile.ai[0];
+							int Type = (int)Projectile.ai[0];
 							Item newItem = new Item();
 							newItem.SetDefaults(Type, true);
 							if(Type == 3196)
```
Comment (gardenappl) [inline Projectiles/AutofisherBobber.cs]: Might be fun to see the `ItemID` here.

File: Tiles/Autofisher.cs [hunk 5]
```diff
@@ -75,7 +74,7 @@ public override bool NewRightClick(int i, int j)
 			catch(Exception e)
 			{
 				GoldensMisc.Log(e);
-				Main.NewText("Error! Report this error and send the Terraria/Logs/Logs.txt file!");
+				Main.NewText("Error! Report this error and send the Terraria/Modloader/Logs/Client.log file!");
```
Comment (gardenappl) [inline Tiles/Autofisher.cs:77]: Good catch.

File: Tiles/ChestVacuum.cs [hunk 6]
```diff
@@ -45,21 +44,35 @@ public void SetDefaultsPostContent()
...
 			tileObjectData.AnchorAlternateTiles = anchorTileList.ToArray();
 		}
 
-		public override void KillMultiTile(int i, int j, int frameX, int frameY)
+		public override bool CanPlace(int i, int j)
 		{
-			Item.NewItem(i * 16, j * 16, 48, 32, mod.ItemType(GetType().Name));
+			Tile tile1 = Main.tile[i, j + 1];
+			Tile tile2 = Main.tile[i + 1, j + 1];
+			if ((tile2 != null || tile2.IsActive) && tile1.frameX != (tile2.frameX - 18))
```
Comment (gardenappl) [inline Tiles/ChestVacuum.cs:59]: Seems like a null reference could happen here?

File: description.txt [hunk 7]
```diff
@@ -13,4 +13,4 @@ Current features include:
  
 And a lot more! Go to the forum page for the full list.
  
-Thanks to Zerokk for his help with sprites and to Pop000100 for maintaining the mod.
+Thanks to Zerokk for his help with sprites and to Pop000100 for maintaining the Mod.
```
Comment (gardenappl) [inline description.txt]: Why the uppercase?

File: Projectiles/AutofisherBobber.cs (hunk 4 above)
Comment (gardenappl) [reply Projectiles/AutofisherBobber.cs]: On second thought, after looking at the later autofisher code, I think it's okay to just keep the numbers.

**Review by gardenappl — COMMENTED** [review]
Amazing work! I only have a couple of questions.

File: NPCs/BaseballCaveBat.cs (hunk 2 above)
Comment (gardenappl) [reply NPCs/BaseballCaveBat.cs:46]: (forgot this was a rare enemy and not a GlobalNPC for standard bats)

File: Projectiles/AutofisherBobber.cs (hunk 3 above)
Comment (Pop000100) [reply Projectiles/AutofisherBobber.cs]: I'll remove all the commented out code, I had to make changes to the auto fisher again(fishing incorrect items).

File: description.txt (hunk 7 above)
Comment (Pop000100) [reply description.txt]: That is a mistake I used visual studio to replace all "mod." figuring there would be no collateral damage I was barely wrong.

File: Tiles/ChestVacuum.cs (hunk 6 above)
Comment (Pop000100) [reply Tiles/ChestVacuum.cs:59]: Let me check on a server, tiles can't be null on a client(and single player). This isn't true on a server I spaced that out.

Comment (Pop000100) [reply Tiles/ChestVacuum.cs:59]: It looks like this hook is never called on the server so it is safe.

File: Tiles/Autofisher.cs (hunk 5 above)
Comment (Pop000100) [reply Tiles/Autofisher.cs:77]: Thank you

File: Tiles/ChestVacuum.cs (hunk 6 above)
Comment (gardenappl) [reply Tiles/ChestVacuum.cs:59]: I just meant like
`tile2 != null && tile2.IsActive` 
rather than
`tile2 != null || tile2.IsActive`

**Comment by gardenappl** [comment]
Sorry to bother you, but how hard would it be to allow the Autofisher to fish up enemies? I remember back in 1.3 I mentioned it would be fun if it could summon Duke Fishron.

**Comment by Pop000100** [comment]
the recent update to the fisher should make it easier, I'll look into it. but it probably should be split from this PR.

I thought about doing it for the bloodmoon enemies but I decided to not since you asked for me to not add or remove content. And I saw you commented out Duke Fishron.

Did you just decide against it or did it have problems?

**Comment by gardenappl** [comment]
Well, in 1.3 I thought that automatically summoning Duke Fishron would be just a novelty/easter egg, so I said it would be cool but then just didn't bother. (and after first developing the Autofisher I honestly did not want to touch that code again.)

In 1.4 it seems like fishing enemies plays a slightly bigger role, so now it might be worth looking into.

**Comment by Pop000100** [comment]
> and after first developing the Autofisher I honestly did not want to touch that code again.
I am getting there. Still one more thing to do before its done.
> 
> In 1.4 it seems like fishing enemies plays a slightly bigger role, so now it might be worth looking into.

I'll see about adding it since i have to dive back into the Autofisher anyway.

**Comment by Pop000100** [comment]
Added the ability to spawn enemies for the autofisher including the bloodmoon enemies.

**Comment by gardenappl** [comment]
Awesome! I don't know how to thank you for the hard work.

**Comment by Pop000100** [comment]
When you have time can you review and approve the changes.

**Comment by gardenappl** [comment]
I've been reviewing commits as they came along, though mostly skimming a lot of big changes.

The code *looks* fine, and I assume you tested it.

One tiny concern: is it necessary to use .hjson for localization? Do we need to convert the Russian and Chinese files, too? The Russian lang just needs one simple tweak: replace 30% with 35% for Autofisher tooltip.

**Comment by Pop000100** [comment]
I have tested it and I had someone report some bugs and I gave him a pre-release build, he said it fixed his issues. I can a revert the lang files for now, I don't know when or if they will be ending support for .lang.

EDIT: Seems like they removed .lang support completely.

Yes in that case I need to convert both the chinese and russian localizations

There is a bug in the russian file that crashes the converter tool.
`WorldGen.AncientForge=Добавляем древние кузницы...
WorldGen.AncientForge=Добавляем древние адские кузницы...`

Which one is correct?

**Comment by gardenappl** [comment]
Whoops, it should be:
```
WorldGen.AncientForge=Добавляем древние кузницы...
WorldGen.AncientHellforge=Добавляем древние адские кузницы...
```

**Comment by Pop000100** [comment]
I'd like to merge it.
Any final thoughts?

**Comment by gardenappl** [comment]
I think we're good now, thanks! I will update the forum post later today.

**Comment by gardenappl** [comment]
How do mod releases work for 1.4? Is it just in Steam Workshop?
I think maybe you should announce a new release on the forums, and then I'll edit the initial post with updated info.

**Comment by Pop000100** [comment]
Yeah you publish ingame through tmodloader and it pushes to workshop.
Want me to do it?

Edit: I accidentally removed you from contributors trying to see If I could switch roles(I am the owner, you are a contributor, I felt this was inaccurate.)
As a aside do you get pings everytime someone posts on the mods or just me?

Edit 2 : Published the update since I noticed you don't have TModloader installed on steam
I edited my post since I can't double post.

**Comment by Pop000100** [comment]
want me to do it?

**Comment by gardenappl** [comment]
I got a ping when you changed my role in the project, other than that, I get nothing. I don't know when the mod updates are published exactly.

**Pull request merged**